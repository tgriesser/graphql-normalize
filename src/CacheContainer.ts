import produce, { Immutable } from 'immer';
import type { CacheShape } from './cache';

export class CacheContainer {
  #state: Immutable<CacheShape>;
  #subscribers: Function[] = [];

  constructor(initialState?: CacheShape) {
    this.#state = produce({ fields: {}, operations: {} }, (root) => initialState ?? root);
  }

  getState() {
    return this.#state;
  }

  subscribe(fn: (state: Immutable<CacheShape>) => void) {
    this.#subscribers.push(fn);
    return () => {
      this.#subscribers.splice(this.#subscribers.indexOf(fn), 1);
    };
  }
}
