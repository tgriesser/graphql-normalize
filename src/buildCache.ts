import type { FormattedExecutionResult } from 'graphql';
import { walk } from './walk';
import type { NormalizedDoc } from './extensionShape';

/**
 * Write values into the cache, given the values, the existing cache, and the
 */
export function buildCache(result: FormattedExecutionResult, normalizingMeta: NormalizedDoc) {
  let pathPrefix = '';
  let currentKey;
  let currentKeyStack = [];
  let currentWriteKey = undefined;
  let currentWriteKeyStack = [];

  const selectionKeys = Object.keys(normalizingMeta.selectionSet);

  function walkUntilList(val: any) {
    walk(val, {
      enter(node) {
        const key = node.path.join('.');
        const metaDef = normalizingMeta.selectionSet[key];
        console.log(key, metaDef);
        if (metaDef.list) {
          return false;
        }
      },
      leave() {
        //
      },
    });
  }

  walkUntilList(result);
}
