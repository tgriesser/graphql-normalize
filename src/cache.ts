import type { NormalizeMetaShape } from './metadataShapes';

export interface CacheShape {
  operations: {
    // The metadata for each query
    [hash: string]: {
      meta: NormalizeMetaShape;
    };
  };
  fields: {
    [key: string]: any;
  };
}
