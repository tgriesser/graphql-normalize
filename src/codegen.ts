import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
