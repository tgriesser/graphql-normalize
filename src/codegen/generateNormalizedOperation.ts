import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';
import type { TypePolicies } from './getCacheKey';

export function generateNormalizedOperation(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedOperations(schema, [document], typePolicies)[0];
}

/**
 * Given a schema & operation, flattens the fragments and adds any necessary missing fields
 * in order to properly normalize the response in a cache.
 */
export function generateNormalizedOperations(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
) {
  let toNormalize = documents;
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    toNormalize = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    });
  }
  return addNormalizingFields(schema, toNormalize, typePolicies);
}
