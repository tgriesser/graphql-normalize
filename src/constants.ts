export const __typename = '__typename' as const;

export const NO_ARGS = '$';
