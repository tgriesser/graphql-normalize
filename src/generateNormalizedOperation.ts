import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';

export function generateNormalizedOperation(schema: GraphQLSchema, operation: DocumentNode) {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }
  return addNormalizingFields(schema, finalOperation);
}
