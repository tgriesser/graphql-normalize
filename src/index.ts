import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
