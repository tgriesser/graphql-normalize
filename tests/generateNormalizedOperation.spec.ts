import { describe, expect, it } from 'vitest';
import { schema } from './fixtures/schema';
import { generateNormalizedOperation } from '../src/generateNormalizedOperation';
import { printAddedFields } from './utils/printAddedFields';
import { operation1Doc } from './fixtures/ops';

describe('generateNormalizedOperation', () => {
  it('adds the necessary fields to properly normalize', () => {
    const changed = generateNormalizedOperation(schema, operation1Doc);
    expect(printAddedFields(operation1Doc, changed)).toMatchSnapshot();
  });

  //
});
