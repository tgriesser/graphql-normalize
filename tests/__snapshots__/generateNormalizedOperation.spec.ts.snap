// Vitest Snapshot v1

exports[`generateNormalizedOperation > adds the necessary fields to properly normalize 1`] = `
"query Operation1($first: Int = 12, $nodeId: ID! = \\"\\", $hasNode: Boolean = false) {
  posts(first: $first) {
    nodes {
      __typename
      author {
        __typename
      }
    }
    edges {
      cursor
      __typename
      node {
        __typename
        author {
          __typename
        }
      }
    }
  }
  first10Posts: posts(first: 10) {
    nodes {
      __typename
    }
  }
  currentNode: node(id: $nodeId) @include(if: $hasNode) {
    __typename
    ... on User {
      id
    }
    ... on Post {
      __typename
    }
  }
  homeItems(first: 10) {
    nodes {
      __typename
      ... on Blog {
        __typename
      }
    }
  }
}"
`;
